mlflow/models/evaluation/base.py

```diff
@@ -0,0 +1,530 @@
+            return self.data, self.labels
+        else:
+            if not isinstance(self.data, pd.DataFrame):
+                data = self.data.limit(EvaluationDataset.SPARK_DATAFRAME_LIMIT).toPandas()
+                _logger.warning(
+                    f"Only the first {EvaluationDataset.SPARK_DATAFRAME_LIMIT} rows in the "
+                    f"spark dataframe are examined."
+                )
+            else:
+                data = self.data
+            feature_cols = [x for x in data.columns if x != self.labels]
+            return data[feature_cols], data[self.labels]
+
+    @staticmethod
+    def _array_like_obj_to_bytes(data):
+        """
+        Helper method to convert pandas dataframe/numpy array/list into bytes for
+        MD5 calculation purpose.
+        """
+        if isinstance(data, pd.DataFrame):
+            return data.to_numpy().tobytes() + ",".join(list(data.columns)).encode("UTF-8")
+        elif isinstance(data, np.ndarray):
+            return data.tobytes()
+        elif isinstance(data, list):
+            return np.array(data).tobytes()
+        else:
+            raise ValueError("Unsupported data type.")
+
+    @staticmethod
+    def _gen_md5_for_arraylike_obj(md5_gen, data):
+        """
+        Helper method to generate MD5 hash array-like object, the MD5 will calculate over:
+         - array length
+         - first NUM_SAMPLE_ROWS_FOR_HASH rows content
+         - last NUM_SAMPLE_ROWS_FOR_HASH rows content
+        """
+        md5_gen.update(np.int64(len(data)).tobytes())
+        if len(data) < EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH * 2:
+            md5_gen.update(EvaluationDataset._array_like_obj_to_bytes(data))
+        else:
+            md5_gen.update(
+                EvaluationDataset._array_like_obj_to_bytes(
+                    data[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]
+                )
+            )
+            md5_gen.update(
+                EvaluationDataset._array_like_obj_to_bytes(
+                    data[-EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH :]
+                )
+            )
+
+    @property
+    def name(self):
+        """
+        Dataset name, which is specified dataset name or the dataset hash if user don't specify
+        name.
+        """
+        return self._user_specified_name if self._user_specified_name is not None else self.hash
+
+    @property
+    def hash(self):
+        """
+        Compute a hash from the specified dataset by selecting the first 5 records, last 5 records,
+        dataset size and feeding them through a cheap, low-collision hash function
+        """
+        if self._hash is None:
+            md5_gen = hashlib.md5()
+            if isinstance(self.data, np.ndarray):
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.labels)
+            elif isinstance(self.data, pd.DataFrame):
```

Comment:
Can we handle the Spark DataFrame hashing here as well?

Reply:
Good catch!